### ISSUE TITLE
modal: not working with version 2.0.0-rc.20

### ISSUE THREAD

**Issue by xaviarnl** [bootstrap-vue/bootstrap-vue#3312]
### Describe the bug

When we upgrade from rc.19 to rc.20, none of the Bootstrap Vue modals work anymore on our site. The call to "show()" just doesn't do anything: no modal div is added to the DOM, but also there are no messages whatsoever (e.g. warnings or errors) on the error console.
I double checked, and if I downgrade to rc.19, the modals work again, and if I then upgrade again to rc.20 they break again.

### Versions

**Libraries:**

- BootstrapVue: 2.0.0-rc.20
- Bootstrap: 4.3.1
- Vue: 2.6.10

**Environment:**

- Device: PC
- OS: Windows 10
- Browser: Chrome, v74.0.3729.131

**Comment by AM-PRO** [comment]
Same bug here but with "bootstrap-vue": "^2.0.0-rc.19" :(

**Comment by Qvatra** [comment]
same here on version rc.20
I have noticed that it is related to the way you are opening a modal.
v-b-modal directive works fine but v-model and visible prop result in an error:

```
TypeError: Cannot read property 'data' of undefined
    at app.js:1257
    at Array.forEach (<anonymous>)
    at getNuxtChildComponents (app.js:1256)
    at hotReloadAPI (app.js:1272)
    at app.js:1436
    at Array.<anonymous> (commons.app.js:15205)
    at flushCallbacks (commons.app.js:15124) 
```

**Comment by tmorehouse** [comment]
Could you provide a minimal reproduction of the bug? (jsfiddle, codesandbox, codepen)?

How are you opening the modals?  using the `show()` method on a `$ref` to the modal, using the `this.$bvModal.show()`

Is this during dev HMR (not module reload)? What other modules are you using (i.e. are you using PortalVue in your app?)

**Comment by tmorehouse** [comment]
@Qvatra The following is working fine for me:

```html
<template>
  <div>
    <b-button @click="modalShow = !modalShow">Open Modal</b-button>
    <b-modal v-model="modalShow">Hello From Modal!</b-modal>
  </div>
</template>
<script>
  export default {
    data() {
      return {
        modalShow: false
      }
    }
  }
</script>
```

The `visible` prop is part of the `v-model` and should not be used directly (as it needs to be synced with the v-model event)

**Comment by jackmu95** [comment]
Maybe it is related to the new `static` handling and modals are always lazy loaded by default now.

**Comment by tmorehouse** [comment]
Possible... but the `$refs` should always be there, even when lazy-loaded

**Comment by jackmu95** [comment]
@tmorehouse Yeah, `$refs` of the modal itself but not of the content.

**Comment by tmorehouse** [comment]
true, but the OP is refereing to showing the modal via the `show()` method.

**Comment by xaviarnl** [comment]
When I was trying to make the modals work, I tried both options, but neither worked. By default, we are using the show() method on a $ref.

**Comment by tmorehouse** [comment]
@xaviarnl can you provide some sample markup?

**Comment by tmorehouse** [comment]
Does you modal have the ID set?

**Comment by xaviarnl** [comment]
The problems occured in both my local development environment, as well as in our production environment (with a (slightly) different, production build).

We are using all kinds of Vue modules. However, we are not using PortalVue or vue-router (this is not a SPA).

**Comment by xaviarnl** [comment]
> Does you modal have the ID set?

Yes it does. And a ref (obviously).

**Comment by xaviarnl** [comment]
It is hard to provide sample code, as the modals are all pretty deeply buried in our page structure. And our application is not a public one.

**Comment by tmorehouse** [comment]
Are you able to reduce the issue to the bare minimum that can reproduce the issue?

Are the modal's placed below or above the trigger elements in the DOM order?

If you place your show call inside a `$nextTick()` does that make it work?

Do you have [Vue aliased](https://bootstrap-vue.js.org/docs#using-module-bundlers) in your webpack config?

**Comment by tmorehouse** [comment]
@xaviarnl what if you set the `static` prop on `<b-modal>` (which forces the modal to be rendered in place)

**Comment by Qvatra** [comment]
@tmorehouse, I've reproduced the bug with the example you gave.
The only change I did is set default value of modalShow to true:
```
<template>
  <div>
    <b-button @click="modalShow = !modalShow">Open Modal</b-button>
    <b-modal v-model="modalShow">Hello From Modal!</b-modal>
  </div>
</template>
<script>
  export default {
    data() {
      return {
        modalShow: true
      }
    }
  }
</script>
```
This is the exact case that I have in my app - in some cases modal is opened by default

**Comment by xaviarnl** [comment]
It is not possible for me to provide you with a jsfiddle (or similar) in the short term, as I have already lost a lot of hours on this.

We do not use HMR. 
The other modules that we are using, are:
- axios (0.18.0)
- d3 (5.9.2)
- document-register-element (1.13.1)
- jquery (3.3.1)
- schema-utils (1.0.0)
- vue-axios (2.1.4)
- vue-custom-element (3.2.6)
- vue-multiselect (2.1.4)
- vue2-dragula (2.5.4)
- vuejs-datepicker ).9.29)
- vuex (3.0.1)

**Comment by Qvatra** [comment]
> @xaviarnl what if you set the `static` prop on `<b-modal>` (which forces the modal to be rendered in place)

That did the job for me. Until the issue is solved I'm gonna use static + lazy

**Comment by tmorehouse** [comment]
@Qvatra thanks for the repro... We may need to add the `$nextTick` in b-modal when the modal is initially open modal on `mount()`

**Comment by tmorehouse** [comment]
@xaviarnl without knowing what is happening in your app to generate the modals, we are blind as to reproducing your issue.

**Comment by xaviarnl** [comment]
I totally get that a reproducable situation/codepen would be the best option, but I unfortunately do not have the time for that at the moment. So if that means the issue will not be resolved (soon), than that is how it is. I know that for now I have a workable situation by locking down the version to rc.19.

**Comment by tmorehouse** [comment]
@xaviarnl  Or try using the `static` prop

**Comment by h0rn3z0r** [comment]
@tmorehouse, please, take a look on this demo: https://jsfiddle.net/fjk5tcbh/
This is an approximate version of my issue with the modal component after updating to rc20. The input field should be focused after 1 sec delay. Though this can be fixed by adding another nextTick in a watch property or using a static prop. Does it mean that I must use a static prop for this cases from now on or will it be fixed?

**Comment by tmorehouse** [comment]
@h0rn3z0r You shouldn't rely on a delay for auto-focusinig an input in the modal.  Use the `@shown` event to trigger your auto-focus of the input.  This will ensure that teh modal has been fully shown before the auto-focus takes place.

**Comment by tmorehouse** [comment]
@h0rn3z0r We have discovered a caveat with PortalVue and `$refs` (as well as provide and inject hierarchy), which causes any content that is portalled to loose its `$parent` relationship (which can interfere with `$refs`).

We have updated modal to use to our own portalling code (which will be available in 2.0.0-rc.21 early next week). which doesn't have the issue that PortalVue does.

**Comment by h0rn3z0r** [comment]
> 
> 
> @h0rn3z0r You shouldn't rely on a delay for auto-focusinig an input in the modal. Use the `@shown` event to trigger your auto-focus of the input. This will ensure that the modal has been fully shown/rendered before the auto-focus takes place.

@tmorehouse Thanks for the answer, unfortunately, `@shown` method is not an option for me. I have a simple modal with AJAX-generated content (setTimeout was used in the demo for simplification). After receiving the response from backend I replace the spinner with a form and do `focus()` on `$nextTick` according to Vue docs. The `@shown` event will fire long before the input field appears.

I have tested my code in the playground and now the input field shows only on 3rd `$nextTick`. It happens with `document.querySelector('input')` as well, not only with `$refs`. Anyway, using a `static` prop is a suitable solution for me, though this situation with "ticks" is not very intuitive.

**Comment by tmorehouse** [comment]
@h0rn3z0r Here is probably a better way for you to autofocus the input (and not be dependent on the underlying implementation/delays):

https://jsfiddle.net/5phbsz4m/

The above fiddle uses a custom directive `v-autofocus` that focuses the input when the element is _inserted_ into the DOM (works only if using `v-if`, not `v-show`).

The directive code is:

```js
autofocus = {
  inserted(el) {
    el.focus()
  }
}
```

The above directive assumes the root element of the input/component is the input.

For handling edge cases where the input _might be nested_ deeper inside a component, you can do the following which will work for both components where the input is the root element as well as inputs nested inside additional markup (i.e. `b-form-file`, `b-form-checkbox`, `b-form-radio`, etc):

```js
autofocus = {
  inserted(el) {
    const selector = 'input, textarea, select'
    if (!el.matches(selector)) {
      el = el.querySelector(selector)
    }
    el && el.focus &&el.focus()
  }
}
```

**Comment by tmorehouse** [comment]
@h0rn3z0r PR https://github.com/bootstrap-vue/bootstrap-vue/pull/3341 adds in support for auto focusing `b-form-xxx` form components via a new prop `autofocus`

**Comment by h0rn3z0r** [comment]
@tmorehouse Thanks for your work, that is indeed a proper way to do this

**Comment by jackmu95** [comment]
Closed by #3320, #3326 and #3341.

**Comment by irufus** [comment]
Thanks for fixing this one.
